src/Microsoft.VisualStudio.Editor.Razor/DefaultVisualStudioRazorParser.cs

```diff
@@ -0,0 +1,372 @@
+            ICompletionBroker completionBroker,
+            IEditorOperationsFactoryService editorOperationsFactory,
+            IEnumerable<ICanHasContextChangedListener> parserContextChangedListeners)
+        {
+            if (dispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(dispatcher));
+            }
+
+            if (documentTracker == null)
+            {
+                throw new ArgumentNullException(nameof(documentTracker));
+            }
+
+            if (templateEngineFactory == null)
+            {
+                throw new ArgumentNullException(nameof(templateEngineFactory));
+            }
+
+            if (string.IsNullOrEmpty(filePath))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(filePath));
+            }
+
+            if (string.IsNullOrEmpty(projectDirectory))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(projectDirectory));
+            }
+
+            if (errorReporter == null)
+            {
+                throw new ArgumentNullException(nameof(errorReporter));
+            }
+
+            if (completionBroker == null)
+            {
+                throw new ArgumentNullException(nameof(completionBroker));
+            }
+
+            if (editorOperationsFactory == null)
+            {
+                throw new ArgumentNullException(nameof(editorOperationsFactory));
+            }
+
+            if (parserContextChangedListeners == null)
+            {
+                throw new ArgumentNullException(nameof(parserContextChangedListeners));
+            }
+
+            _dispatcher = dispatcher;
+            _templateEngineFactory = templateEngineFactory;
+            _errorReporter = errorReporter;
+            _completionBroker = completionBroker;
+            _parserContextChangedListeners = parserContextChangedListeners;
+            _documentTracker = documentTracker;
+            FilePath = filePath;
+            _projectDirectory = projectDirectory;
+            TextBuffer = _documentTracker.TextBuffer;
+
+            _documentTracker.ContextChanged += DocumentTracker_ContextChanged;
+            TextBuffer.Changed += TextBuffer_OnChanged;
+            _braceSmartIndenter = new BraceSmartIndenter(_dispatcher, _documentTracker, editorOperationsFactory);
+        }
+
+        public override event EventHandler<DocumentStructureChangedEventArgs> DocumentStructureChanged;
+
+        public override RazorTemplateEngine TemplateEngine => _templateEngine;
+
+        public override string FilePath { get; }
+
+        public override RazorCodeDocument CodeDocument => _codeDocument;
+
+        public override ITextSnapshot Snapshot => _snapshot;
+
+        public override ITextBuffer TextBuffer { get; }
+
+        public override async void Reparse()
+        {
+            // Can be called from any thread
```

Comment:
Who is calling this from a background thread?

Reply:
This should return `Task`

Reply:
BTW the problem with this approach is that callers can't look at the results due to threading.

Reply:
> Who is calling this from a background thread?

When Idle determines it needs to be called and when the WTE code determines TagHelpers have changed. Eventually this method wont need to be exposed at all once we own the entire TagHelper decision stack.



> BTW the problem with this approach is that callers can't look at the results due to threading.

Hmm? not sure I follow, could you expand.